This notebook will containg all data cleaning aspects to ensure that our data is propraly formatted and ready to create a machine learning model. 

In [193]:
#Import packages
import pandas as pd

# Player Statistics Data Cleaning

First we need to clean the data for our player statistics

In [194]:
#Load csv file
stats_df=pd.read_csv('stats_df.csv')

In [195]:
stats_df.head(5)

Unnamed: 0  Rk                Player Pos  Age   Tm   G  GS    MP   FG  ...  \
0           0   1  Kareem Abdul-Jabbar*   C   39  LAL  78  78  31.3  7.2  ...   
1           1   2           Alvan Adams   C   32  PHO  68  40  24.9  4.6  ...   
2           2   3         Michael Adams  PG   24  WSB  63   0  20.7  2.5  ...   
3           3   4        Rafael Addison  SF   22  PHO  62  12  11.5  2.4  ...   
4           4   5          Mark Aguirre  SF   27  DAL  80  80  33.3  9.8  ...   

   ORB  DRB  TRB  AST  STL  BLK  TOV   PF   PTS  Year  
0  1.9  4.8  6.7  2.6  0.6  1.2  2.4  3.1  17.5  1987  
1  1.3  3.6  5.0  3.3  0.9  0.5  2.0  3.0  11.1  1987  
2  0.6  1.3  2.0  3.9  1.3  0.1  1.3  1.4   7.2  1987  
3  0.7  1.0  1.7  0.7  0.4  0.1  0.9  1.2   5.8  1987  
4  2.3  3.1  5.3  3.2  1.1  0.4  2.7  3.0  25.7  1987  

[5 rows x 32 columns]

First we will remove the Unamed: 0 and Rk column, as we do not need them.

In [196]:
del stats_df['Unnamed: 0']
del stats_df['Rk']

In [197]:
stats_df.head(5)

Player Pos  Age   Tm   G  GS    MP   FG   FGA    FG%  ...  \
0  Kareem Abdul-Jabbar*   C   39  LAL  78  78  31.3  7.2  12.7  0.564  ...   
1           Alvan Adams   C   32  PHO  68  40  24.9  4.6   9.1  0.503  ...   
2         Michael Adams  PG   24  WSB  63   0  20.7  2.5   6.2  0.407  ...   
3        Rafael Addison  SF   22  PHO  62  12  11.5  2.4   5.3  0.441  ...   
4          Mark Aguirre  SF   27  DAL  80  80  33.3  9.8  19.9  0.495  ...   

   ORB  DRB  TRB  AST  STL  BLK  TOV   PF   PTS  Year  
0  1.9  4.8  6.7  2.6  0.6  1.2  2.4  3.1  17.5  1987  
1  1.3  3.6  5.0  3.3  0.9  0.5  2.0  3.0  11.1  1987  
2  0.6  1.3  2.0  3.9  1.3  0.1  1.3  1.4   7.2  1987  
3  0.7  1.0  1.7  0.7  0.4  0.1  0.9  1.2   5.8  1987  
4  2.3  3.1  5.3  3.2  1.1  0.4  2.7  3.0  25.7  1987  

[5 rows x 30 columns]

You may have noticed that Kareem Abdul-Jabbar has an * next to his name. This means that he was an all-star for the season. However, since we want to merge this we our MVP data frame we need to remove this so we can merge the data. Sorry Kareem.

In [198]:
#Remove *
stats_df['Player'] = stats_df['Player'].str.rstrip('*')

In [199]:
stats_df.head(5)

Player Pos  Age   Tm   G  GS    MP   FG   FGA    FG%  ...  \
0  Kareem Abdul-Jabbar   C   39  LAL  78  78  31.3  7.2  12.7  0.564  ...   
1          Alvan Adams   C   32  PHO  68  40  24.9  4.6   9.1  0.503  ...   
2        Michael Adams  PG   24  WSB  63   0  20.7  2.5   6.2  0.407  ...   
3       Rafael Addison  SF   22  PHO  62  12  11.5  2.4   5.3  0.441  ...   
4         Mark Aguirre  SF   27  DAL  80  80  33.3  9.8  19.9  0.495  ...   

   ORB  DRB  TRB  AST  STL  BLK  TOV   PF   PTS  Year  
0  1.9  4.8  6.7  2.6  0.6  1.2  2.4  3.1  17.5  1987  
1  1.3  3.6  5.0  3.3  0.9  0.5  2.0  3.0  11.1  1987  
2  0.6  1.3  2.0  3.9  1.3  0.1  1.3  1.4   7.2  1987  
3  0.7  1.0  1.7  0.7  0.4  0.1  0.9  1.2   5.8  1987  
4  2.3  3.1  5.3  3.2  1.1  0.4  2.7  3.0  25.7  1987  

[5 rows x 30 columns]

The asterisk is now gone!

If you explored the data on the basketball reference website you may have seen that some players names are listed twice in a given year. This is because they played for more than one team during that season. We need to fix this so they are only listed once. In case you did not look at the basketball reference website. I've defined a function to show that some players appear twice.

In [200]:
def player_count_year(df, year):
    #Filter data frame
    filtered_df = df[df['Year'] == year]
    # Count the occurrences of each player
    player_counts = filtered_df['Player'].value_counts()
    
    return player_counts


In [201]:
#Call the function on a year
player_count_year(stats_df, 1987)

Chris Engler        4
Perry Young         3
Steve Colter        3
Tyrone Corbin       3
Earl Cureton        3
                   ..
Roy Hinson          1
Rod Higgins         1
Kevin Henderson     1
Jerome Henderson    1
Bill Martin         1
Name: Player, Length: 334, dtype: int64

We can see that there are multiple players that appear twice (or more) in the data set. First let's look closer at Chris Engler playing for three teams in 1987.

In [202]:
stats_df.groupby(['Player', 'Year']).get_group(('Chris Engler', 1987))

Player Pos  Age   Tm   G  GS   MP   FG  FGA    FG%  ...  ORB  DRB  \
95  Chris Engler   C   27  TOT  30   0  6.5  0.8  1.7  0.451  ...  0.8  1.1   
96  Chris Engler   C   27  POR   7   0  2.4  0.6  1.1  0.500  ...  0.9  0.3   
97  Chris Engler   C   27  MIL   5   0  9.6  0.6  2.4  0.250  ...  0.6  2.6   
98  Chris Engler   C   27  NJN  18   0  7.2  0.9  1.7  0.516  ...  0.8  1.1   

    TRB  AST  STL  BLK  TOV   PF  PTS  Year  
95  1.9  0.3  0.2  0.4  0.4  1.1  1.9  1987  
96  1.1  0.1  0.1  0.1  0.3  0.0  1.6  1987  
97  3.2  0.6  0.2  0.2  0.0  2.0  1.4  1987  
98  1.8  0.2  0.2  0.5  0.6  1.3  2.2  1987  

[4 rows x 30 columns]

To deal with this we a will define a function the converts players with multiple rows, like Chris Engler, to a single row. Luckily, basketball reference has a row called TOT which is their total stats for the season so we do not have to deal with calculating this ourselves. TOT is not a valid NBA team so we will replace TOT with the name of the players most recent team for the season.

In [203]:
def one_team(df):
    return df if df.shape[0] == 1 else df[df["Tm"].eq("TOT")].assign(Tm=df.iloc[-1]["Tm"])

stats_df = stats_df.groupby(["Player", "Year"]).apply(one_team)

In [204]:
stats_df.head()

Player Pos  Age   Tm   G  GS    MP   FG  FGA    FG%  \
Player     Year                                                                 
A.C. Green 1987 118   A.C. Green  PF   23  LAL  79  72  28.4  4.0  7.4  0.538   
           1988 504   A.C. Green  PF   24  LAL  82  64  32.1  3.9  7.8  0.503   
           1989 938   A.C. Green  PF   25  LAL  82  82  30.6  4.9  9.2  0.529   
           1990 1383  A.C. Green  PF   26  LAL  82  82  33.0  4.7  9.8  0.478   
           1991 1860  A.C. Green  PF   27  LAL  82  21  26.4  3.1  6.6  0.476   

                      ...  ORB  DRB  TRB  AST  STL  BLK  TOV   PF   PTS  Year  
Player     Year       ...                                                      
A.C. Green 1987 118   ...  2.7  5.1  7.8  1.1  0.9  1.0  1.3  2.2  10.8  1987  
           1988 504   ...  3.0  5.7  8.7  1.1  1.1  0.5  1.5  2.5  11.4  1988  
           1989 938   ...  3.1  5.9  9.0  1.3  1.1  0.7  1.5  2.1  13.3  1989  
           1990 1383  ...  3.2  5.5  8.7  1.1  0.8  0.6  1.4  2.5  12.9  1990  
           1991 1860  ...  2.5  3.8  6.3  0.9  0.7  0.3  1.2  1.4   9.1  1991  

[5 rows x 30 columns]

In [205]:
#Check to see if the function worked
player_count_year(stats_df, 1987)

A.C. Green          1
Mike Glenn          1
Mike Brown          1
Mike Brittain       1
Michael Wilson      1
                   ..
Gene Banks          1
Fred Roberts        1
Fred Cofield        1
Franklin Edwards    1
Xavier McDaniel     1
Name: Player, Length: 333, dtype: int64

Since every player is only listed once now, we can see that our function worked. However, we can see that the indexes of our dataframe have changed. we need to correct this.

In [206]:
#Fix the indexes
stats_df.index = stats_df.index.droplevel()
stats_df.index = stats_df.index.droplevel()

In [207]:
stats_df.head(5)

Player Pos  Age   Tm   G  GS    MP   FG  FGA    FG%  ...  ORB  DRB  \
118   A.C. Green  PF   23  LAL  79  72  28.4  4.0  7.4  0.538  ...  2.7  5.1   
504   A.C. Green  PF   24  LAL  82  64  32.1  3.9  7.8  0.503  ...  3.0  5.7   
938   A.C. Green  PF   25  LAL  82  82  30.6  4.9  9.2  0.529  ...  3.1  5.9   
1383  A.C. Green  PF   26  LAL  82  82  33.0  4.7  9.8  0.478  ...  3.2  5.5   
1860  A.C. Green  PF   27  LAL  82  21  26.4  3.1  6.6  0.476  ...  2.5  3.8   

      TRB  AST  STL  BLK  TOV   PF   PTS  Year  
118   7.8  1.1  0.9  1.0  1.3  2.2  10.8  1987  
504   8.7  1.1  1.1  0.5  1.5  2.5  11.4  1988  
938   9.0  1.3  1.1  0.7  1.5  2.1  13.3  1989  
1383  8.7  1.1  0.8  0.6  1.4  2.5  12.9  1990  
1860  6.3  0.9  0.7  0.3  1.2  1.4   9.1  1991  

[5 rows x 30 columns]

# MVP Data Cleaning

Now, we will clearn the data for the mvp_df file

In [208]:
#Load csv file
mvp_df=pd.read_csv('mvp_df.csv')

In [209]:
mvp_df.head(5)

Unnamed: 0 Rank             Player  Age   Tm  First  Pts Won  Pts Max  \
0           0    1      Magic Johnson   27  LAL   65.0    733.0      780   
1           1    2     Michael Jordan   23  CHI   10.0    449.0      780   
2           2    3         Larry Bird   30  BOS    1.0    271.0      780   
3           3    4       Kevin McHale   29  BOS    0.0    254.0      780   
4           4    5  Dominique Wilkins   27  ATL    0.0    128.0      780   

   Share   G  ...  TRB   AST  STL  BLK    FG%    3P%    FT%    WS  WS/48  Year  
0  0.940  80  ...  6.3  12.2  1.7  0.5  0.522  0.205  0.848  15.9  0.263  1987  
1  0.576  82  ...  5.2   4.6  2.9  1.5  0.482  0.182  0.857  16.9  0.247  1987  
2  0.347  74  ...  9.2   7.6  1.8  0.9  0.525  0.400  0.910  15.2  0.243  1987  
3  0.326  77  ...  9.9   2.6  0.5  2.2  0.604  0.000  0.836  14.8  0.232  1987  
4  0.164  79  ...  6.3   3.3  1.5  0.6  0.463  0.292  0.818  12.2  0.197  1987  

[5 rows x 22 columns]

There are some columns here that we will need to get rid of. For example, AST, STL, BLK and so on, are all apart of the player stats dataset so we do not need them here. Furthermore, the Unnamed: 0 column is not needed.

In [210]:
#Select the columns of interest out of mvp_df
mvp_df = mvp_df[['Player', 'Year', 'Pts Won', 'Pts Max', 'Share']]

In [211]:
mvp_df.head(5)

Player  Year  Pts Won  Pts Max  Share
0      Magic Johnson  1987    733.0      780  0.940
1     Michael Jordan  1987    449.0      780  0.576
2         Larry Bird  1987    271.0      780  0.347
3       Kevin McHale  1987    254.0      780  0.326
4  Dominique Wilkins  1987    128.0      780  0.164

Now our dataframe is reduced to just show the key columns. Let's merge this dataframe with the stats_df.

In [212]:
merged_df = stats_df.merge(mvp_df, how='outer', on=['Player', 'Year'])

In [213]:
merged_df.head(5)

Player Pos  Age   Tm   G  GS    MP   FG  FGA    FG%  ...  AST  STL  \
0  A.C. Green  PF   23  LAL  79  72  28.4  4.0  7.4  0.538  ...  1.1  0.9   
1  A.C. Green  PF   24  LAL  82  64  32.1  3.9  7.8  0.503  ...  1.1  1.1   
2  A.C. Green  PF   25  LAL  82  82  30.6  4.9  9.2  0.529  ...  1.3  1.1   
3  A.C. Green  PF   26  LAL  82  82  33.0  4.7  9.8  0.478  ...  1.1  0.8   
4  A.C. Green  PF   27  LAL  82  21  26.4  3.1  6.6  0.476  ...  0.9  0.7   

   BLK  TOV   PF   PTS  Year  Pts Won  Pts Max  Share  
0  1.0  1.3  2.2  10.8  1987      NaN      NaN    NaN  
1  0.5  1.5  2.5  11.4  1988      NaN      NaN    NaN  
2  0.7  1.5  2.1  13.3  1989      NaN      NaN    NaN  
3  0.6  1.4  2.5  12.9  1990      NaN      NaN    NaN  
4  0.3  1.2  1.4   9.1  1991      NaN      NaN    NaN  

[5 rows x 33 columns]

Looks good! Lets just check the data on a player who recieved NBA votes in their career.

In [214]:
merged_df[merged_df['Player'] == 'LeBron James'].head(5)

Player Pos  Age   Tm   G  GS    MP    FG   FGA    FG%  ...  AST  \
9928  LeBron James  SG   19  CLE  79  79  39.5   7.9  18.9  0.417  ...  5.9   
9929  LeBron James  SF   20  CLE  80  80  42.4   9.9  21.1  0.472  ...  7.2   
9930  LeBron James  SF   21  CLE  79  79  42.5  11.1  23.1  0.480  ...  6.6   
9931  LeBron James  SF   22  CLE  78  78  40.9   9.9  20.8  0.476  ...  6.0   
9932  LeBron James  SF   23  CLE  75  74  40.4  10.6  21.9  0.484  ...  7.2   

      STL  BLK  TOV   PF   PTS  Year  Pts Won  Pts Max  Share  
9928  1.6  0.7  3.5  1.9  20.9  2004     11.0   1230.0  0.009  
9929  2.2  0.7  3.3  1.8  27.2  2005     93.0   1270.0  0.073  
9930  1.6  0.8  3.3  2.3  31.4  2006    688.0   1250.0  0.550  
9931  1.6  0.7  3.2  2.2  27.3  2007    183.0   1290.0  0.142  
9932  1.8  1.1  3.4  2.2  30.0  2008    438.0   1260.0  0.348  

[5 rows x 33 columns]

Nice, looks like it worked for players who recieved votes. We need to make it so the NaN values are 0 in Pts Won and Share because this means the player recieved no votes. We also need to make the Pts Max equal to the given max for that year.

In [215]:
merged_df[['Pts Won', 'Share']] = merged_df[['Pts Won', 'Share']].fillna(0)

In [216]:
merged_df.head(5)

Player Pos  Age   Tm   G  GS    MP   FG  FGA    FG%  ...  AST  STL  \
0  A.C. Green  PF   23  LAL  79  72  28.4  4.0  7.4  0.538  ...  1.1  0.9   
1  A.C. Green  PF   24  LAL  82  64  32.1  3.9  7.8  0.503  ...  1.1  1.1   
2  A.C. Green  PF   25  LAL  82  82  30.6  4.9  9.2  0.529  ...  1.3  1.1   
3  A.C. Green  PF   26  LAL  82  82  33.0  4.7  9.8  0.478  ...  1.1  0.8   
4  A.C. Green  PF   27  LAL  82  21  26.4  3.1  6.6  0.476  ...  0.9  0.7   

   BLK  TOV   PF   PTS  Year  Pts Won  Pts Max  Share  
0  1.0  1.3  2.2  10.8  1987      0.0      NaN    0.0  
1  0.5  1.5  2.5  11.4  1988      0.0      NaN    0.0  
2  0.7  1.5  2.1  13.3  1989      0.0      NaN    0.0  
3  0.6  1.4  2.5  12.9  1990      0.0      NaN    0.0  
4  0.3  1.2  1.4   9.1  1991      0.0      NaN    0.0  

[5 rows x 33 columns]

Now we need to fill the Pts Max column. The Pts Max is different in each year so we need to find its value.

In [217]:
max_pts_by_year = merged_df.groupby('Year')['Pts Max'].max()
max_pts_by_year = pd.DataFrame(max_pts_by_year)
max_pts_by_year = max_pts_by_year.reset_index()
max_pts_by_year.head(5)

Year  Pts Max
0  1987    780.0
1  1988    800.0
2  1989    850.0
3  1990    920.0
4  1991    960.0

Now we need to fill the NaN values with the corresponding Pts Max

In [218]:
#Fill NaN values
merged_df['Pts Max'] = merged_df['Pts Max'].combine_first(merged_df['Year'].map(max_pts_by_year.set_index('Year')['Pts Max']))

In [219]:
merged_df.head(5)

Player Pos  Age   Tm   G  GS    MP   FG  FGA    FG%  ...  AST  STL  \
0  A.C. Green  PF   23  LAL  79  72  28.4  4.0  7.4  0.538  ...  1.1  0.9   
1  A.C. Green  PF   24  LAL  82  64  32.1  3.9  7.8  0.503  ...  1.1  1.1   
2  A.C. Green  PF   25  LAL  82  82  30.6  4.9  9.2  0.529  ...  1.3  1.1   
3  A.C. Green  PF   26  LAL  82  82  33.0  4.7  9.8  0.478  ...  1.1  0.8   
4  A.C. Green  PF   27  LAL  82  21  26.4  3.1  6.6  0.476  ...  0.9  0.7   

   BLK  TOV   PF   PTS  Year  Pts Won  Pts Max  Share  
0  1.0  1.3  2.2  10.8  1987      0.0    780.0    0.0  
1  0.5  1.5  2.5  11.4  1988      0.0    800.0    0.0  
2  0.7  1.5  2.1  13.3  1989      0.0    850.0    0.0  
3  0.6  1.4  2.5  12.9  1990      0.0    920.0    0.0  
4  0.3  1.2  1.4   9.1  1991      0.0    960.0    0.0  

[5 rows x 33 columns]

Now we can clean the team_record file

# Team Record Data Cleaning

In [220]:
teams_df = pd.read_csv('team_record_df.csv')

In [221]:
teams_df.tail(15)

Unnamed: 0   Rk                    Team  Overall   Home   Road  \
1083          17   18         Toronto Raptors    41-41  27-14  14-27   
1084          18   19           Chicago Bulls    40-42  22-19  18-23   
1085          19   20   Oklahoma City Thunder    40-42  24-17  16-25   
1086          20  NaN                     NaN      NaN  Place  Place   
1087          21   Rk                    Team  Overall   Home   Road   
1088          22   21        Dallas Mavericks    38-44  23-18  15-26   
1089          23   22               Utah Jazz    37-45  23-18  14-27   
1090          24   23          Indiana Pacers    35-47  20-21  15-26   
1091          25   24      Washington Wizards    35-47  19-22  16-25   
1092          26   25           Orlando Magic    34-48  20-21  14-27   
1093          27   26  Portland Trail Blazers    33-49  17-24  16-25   
1094          28   27       Charlotte Hornets    27-55  13-28  14-27   
1095          29   28         Houston Rockets    22-60  14-27   8-33   
1096          30   29       San Antonio Spurs    22-60  14-27   8-33   
1097          31   30         Detroit Pistons    17-65   9-32   8-33   

               E           W         A         C  ...    Feb    Mar    Apr  \
1083       26-26       15-15      4-12      10-8  ...    8-3    7-7    3-2   
1084       27-25       13-17       9-9       7-9  ...    5-7    9-6    3-2   
1085       15-15       25-27       6-4       5-5  ...    4-7   10-7    2-2   
1086  Conference  Conference  Division  Division  ...  Month  Month  Month   
1087           E           W         A         C  ...    Feb    Mar    Apr   
1088       10-20       28-24       6-4       2-8  ...    5-6    5-9    1-4   
1089       13-17       24-28       3-7       4-6  ...    5-6    5-9    1-4   
1090       24-28       11-19      7-11       7-9  ...    4-7    6-9    1-3   
1091       21-31       14-16      5-13      8-10  ...    5-6   5-11    1-4   
1092       20-32       14-16      8-10      5-13  ...    6-5    7-8    1-4   
1093       10-20       23-29       1-9       3-7  ...    5-6   3-13    1-4   
1094       15-37       12-18      2-16      6-12  ...    5-6    6-9    1-3   
1095       10-20       12-40       2-8       4-6  ...   1-10   6-11    3-1   
1096       12-18       10-42       3-7       6-4  ...   1-10   4-11    3-2   
1097        8-44        9-21      1-17      2-14  ...    2-8   1-14    1-4   

      Year  May        SE        NW        SW  Jul  Aug  
1083  2023  NaN      12-6       5-5       7-3  NaN  NaN  
1084  2023  NaN      11-7       6-4       5-5  NaN  NaN  
1085  2023  NaN       4-6       9-7      10-8  NaN  NaN  
1086  2023  NaN  Division  Division  Division  NaN  NaN  
1087  2023  NaN        SE        NW        SW  NaN  NaN  
1088  2023  NaN       2-8       9-9       9-7  NaN  NaN  
1089  2023  NaN       6-4      6-10      10-8  NaN  NaN  
1090  2023  NaN      10-8       2-8       4-6  NaN  NaN  
1091  2023  NaN       8-8       4-6       7-3  NaN  NaN  
1092  2023  NaN       7-9       4-6       5-5  NaN  NaN  
1093  2023  NaN       6-4       7-9      10-8  NaN  NaN  
1094  2023  NaN       7-9       4-6       5-5  NaN  NaN  
1095  2023  NaN       4-6      5-13      4-12  NaN  NaN  
1096  2023  NaN       3-7      6-12      3-13  NaN  NaN  
1097  2023  NaN      5-13       5-5       2-8  NaN  NaN  

[15 rows x 30 columns]

You can see that we have some rows where the data is messed up, like in row 1086 and 1087. We need to delete these rows.

In [222]:
#Drop the rows with NaN values for the Overall column
teams_df = teams_df.dropna(subset=['Overall'])
#Drop the rows that don't contain '-' as this is used for all of the record
teams_df = teams_df[teams_df['Overall'].str.contains('-', na=False)]

In [223]:
teams_df.tail(15)

Unnamed: 0  Rk                    Team Overall   Home   Road      E  \
1081          15  16    New Orleans Pelicans   42-40  27-14  15-26  13-17   
1082          16  17           Atlanta Hawks   41-41  24-17  17-24  26-26   
1083          17  18         Toronto Raptors   41-41  27-14  14-27  26-26   
1084          18  19           Chicago Bulls   40-42  22-19  18-23  27-25   
1085          19  20   Oklahoma City Thunder   40-42  24-17  16-25  15-15   
1088          22  21        Dallas Mavericks   38-44  23-18  15-26  10-20   
1089          23  22               Utah Jazz   37-45  23-18  14-27  13-17   
1090          24  23          Indiana Pacers   35-47  20-21  15-26  24-28   
1091          25  24      Washington Wizards   35-47  19-22  16-25  21-31   
1092          26  25           Orlando Magic   34-48  20-21  14-27  20-32   
1093          27  26  Portland Trail Blazers   33-49  17-24  16-25  10-20   
1094          28  27       Charlotte Hornets   27-55  13-28  14-27  15-37   
1095          29  28         Houston Rockets   22-60  14-27   8-33  10-20   
1096          30  29       San Antonio Spurs   22-60  14-27   8-33  12-18   
1097          31  30         Detroit Pistons   17-65   9-32   8-33   8-44   

          W     A     C  ...   Feb   Mar  Apr  Year  May    SE    NW    SW  \
1081  29-23   4-6   5-5  ...   4-6   9-6  3-2  2023  NaN   4-6   9-9  11-5   
1082  15-15  6-12  12-6  ...   6-5   7-8  3-2  2023  NaN   8-8   4-6   5-5   
1083  15-15  4-12  10-8  ...   8-3   7-7  3-2  2023  NaN  12-6   5-5   7-3   
1084  13-17   9-9   7-9  ...   5-7   9-6  3-2  2023  NaN  11-7   6-4   5-5   
1085  25-27   6-4   5-5  ...   4-7  10-7  2-2  2023  NaN   4-6   9-7  10-8   
1088  28-24   6-4   2-8  ...   5-6   5-9  1-4  2023  NaN   2-8   9-9   9-7   
1089  24-28   3-7   4-6  ...   5-6   5-9  1-4  2023  NaN   6-4  6-10  10-8   
1090  11-19  7-11   7-9  ...   4-7   6-9  1-3  2023  NaN  10-8   2-8   4-6   
1091  14-16  5-13  8-10  ...   5-6  5-11  1-4  2023  NaN   8-8   4-6   7-3   
1092  14-16  8-10  5-13  ...   6-5   7-8  1-4  2023  NaN   7-9   4-6   5-5   
1093  23-29   1-9   3-7  ...   5-6  3-13  1-4  2023  NaN   6-4   7-9  10-8   
1094  12-18  2-16  6-12  ...   5-6   6-9  1-3  2023  NaN   7-9   4-6   5-5   
1095  12-40   2-8   4-6  ...  1-10  6-11  3-1  2023  NaN   4-6  5-13  4-12   
1096  10-42   3-7   6-4  ...  1-10  4-11  3-2  2023  NaN   3-7  6-12  3-13   
1097   9-21  1-17  2-14  ...   2-8  1-14  1-4  2023  NaN  5-13   5-5   2-8   

      Jul  Aug  
1081  NaN  NaN  
1082  NaN  NaN  
1083  NaN  NaN  
1084  NaN  NaN  
1085  NaN  NaN  
1088  NaN  NaN  
1089  NaN  NaN  
1090  NaN  NaN  
1091  NaN  NaN  
1092  NaN  NaN  
1093  NaN  NaN  
1094  NaN  NaN  
1095  NaN  NaN  
1096  NaN  NaN  
1097  NaN  NaN  

[15 rows x 30 columns]

Much better!

In [237]:
#Keep only the columns of interests
teams_df = teams_df[['Team', 'Overall', 'Year']]

In [238]:
teams_df.head(5)

Team Overall  Year
0  Los Angeles Lakers   65-17  1987
1      Boston Celtics   59-23  1987
2       Atlanta Hawks   57-25  1987
3    Dallas Mavericks   55-27  1987
4     Detroit Pistons   52-30  1987

Now we need to deveop a colums for wins, losses, and win percentage. We will use the Overall column to do this.

In [239]:
#Wins column
teams_df['Wins'] = teams_df['Overall'].apply(lambda x: int(x.split('-')[0]))
#Losses column
teams_df['Losses'] = teams_df['Overall'].apply(lambda x: int(x.split('-')[1]))
#Win percentage column
teams_df['W%'] = round(teams_df['Wins'] / (teams_df['Wins'] + teams_df['Losses']), 5)

In [240]:
teams_df.head(5)

Team Overall  Year  Wins  Losses       W%
0  Los Angeles Lakers   65-17  1987    65      17  0.79268
1      Boston Celtics   59-23  1987    59      23  0.71951
2       Atlanta Hawks   57-25  1987    57      25  0.69512
3    Dallas Mavericks   55-27  1987    55      27  0.67073
4     Detroit Pistons   52-30  1987    52      30  0.63415

The team names in this data frame do not match with those in the merged_df. We can see this below.

In [241]:
teams_df['Team'].value_counts()

Los Angeles Lakers                   37
Indiana Pacers                       37
Los Angeles Clippers                 37
New York Knicks                      37
San Antonio Spurs                    37
Boston Celtics                       37
Cleveland Cavaliers                  37
Phoenix Suns                         37
Denver Nuggets                       37
Chicago Bulls                        37
Sacramento Kings                     37
Dallas Mavericks                     37
Atlanta Hawks                        37
Golden State Warriors                37
Houston Rockets                      37
Utah Jazz                            37
Philadelphia 76ers                   37
Portland Trail Blazers               37
Milwaukee Bucks                      37
Detroit Pistons                      37
Miami Heat                           35
Minnesota Timberwolves               34
Orlando Magic                        34
Toronto Raptors                      28
New Jersey Nets                      26


In [242]:
merged_df['Tm'].value_counts()

LAC    597
CLE    591
PHI    591
DAL    587
GSW    586
ATL    581
SAS    580
MIL    580
HOU    575
BOS    574
DEN    573
IND    567
LAL    566
NYK    562
POR    562
SAC    562
CHI    562
DET    561
PHO    560
MIA    556
UTA    556
ORL    526
MIN    513
TOR    460
WAS    431
NJN    407
MEM    371
SEA    316
OKC    256
CHH    209
BRK    202
NOP    179
WSB    163
CHA    157
CHO    147
NOH    143
VAN     88
NOK     32
Name: Tm, dtype: int64

This a problem because we are going to want to match these values so we will need to find a way to correct this. To do this I will use a file that cointains the name of NBA teams and their common abbreviations. Both dataframes have the same number of unique items so it is likely that we have an abbreviation for each team name.

In [243]:
#Read file
nicknames = pd.read_csv('nicknames.csv')

In [244]:
nicknames.head(5)

Abbreviation,Name
0      ATL,Atlanta Hawks
1      BRK,Brooklyn Nets
2      BKN,Brooklyn Nets
3     BOS,Boston Celtics
4  CHA,Charlotte Bobcats

In [245]:
nicknames[['Abbreviation', 'Name']] = nicknames['Abbreviation,Name'].str.split(',', n=1, expand=True)
del nicknames['Abbreviation,Name']

In [246]:
nicknames.head(5)

Abbreviation               Name
0          ATL      Atlanta Hawks
1          BRK      Brooklyn Nets
2          BKN      Brooklyn Nets
3          BOS     Boston Celtics
4          CHA  Charlotte Bobcats

Now we can use this to add full team name to our merged dataframe.

In [247]:
#Map
merged_df['Team'] = merged_df['Tm'].map(nicknames.set_index('Abbreviation')['Name'])

In [248]:
merged_df.head(10)

Player Pos  Age   Tm   G  GS    MP   FG   FGA    FG%  ...  STL  BLK  \
0  A.C. Green  PF   23  LAL  79  72  28.4  4.0   7.4  0.538  ...  0.9  1.0   
1  A.C. Green  PF   24  LAL  82  64  32.1  3.9   7.8  0.503  ...  1.1  0.5   
2  A.C. Green  PF   25  LAL  82  82  30.6  4.9   9.2  0.529  ...  1.1  0.7   
3  A.C. Green  PF   26  LAL  82  82  33.0  4.7   9.8  0.478  ...  0.8  0.6   
4  A.C. Green  PF   27  LAL  82  21  26.4  3.1   6.6  0.476  ...  0.7  0.3   
5  A.C. Green  PF   28  LAL  82  53  35.4  4.7   9.8  0.476  ...  1.1  0.4   
6  A.C. Green  PF   29  LAL  82  55  34.4  4.6   8.6  0.537  ...  1.1  0.5   
7  A.C. Green  PF   30  PHO  82  55  34.5  5.7  11.3  0.502  ...  0.9  0.5   
8  A.C. Green  SF   31  PHO  82  52  32.8  3.8   7.5  0.504  ...  0.7  0.4   
9  A.C. Green  SF   32  PHO  82  36  25.8  2.6   5.4  0.484  ...  0.5  0.3   

   TOV   PF   PTS  Year  Pts Won  Pts Max  Share                Team  
0  1.3  2.2  10.8  1987      0.0    780.0    0.0  Los Angeles Lakers  
1  1.5  2.5  11.4  1988      0.0    800.0    0.0  Los Angeles Lakers  
2  1.5  2.1  13.3  1989      0.0    850.0    0.0  Los Angeles Lakers  
3  1.4  2.5  12.9  1990      0.0    920.0    0.0  Los Angeles Lakers  
4  1.2  1.4   9.1  1991      0.0    960.0    0.0  Los Angeles Lakers  
5  1.4  1.7  13.6  1992      0.0    960.0    0.0  Los Angeles Lakers  
6  1.4  1.8  12.8  1993      0.0    980.0    0.0  Los Angeles Lakers  
7  1.2  1.7  14.7  1994      0.0   1010.0    0.0        Phoenix Suns  
8  1.4  1.8  11.2  1995      0.0   1050.0    0.0        Phoenix Suns  
9  1.0  1.7   7.5  1996      0.0   1130.0    0.0        Phoenix Suns  

[10 rows x 34 columns]

Finally, we need to combines out team_df with merged_df to get a final dataframe containing all stats.

In [250]:
#Merge to all_stats
all_stats = merged_df.merge(teams_df, how='outer', on=['Team', 'Year'])

In [253]:
all_stats.head(5)

Player Pos  Age   Tm   G  GS    MP   FG   FGA    FG%  ...   PTS  \
0      A.C. Green  PF   23  LAL  79  72  28.4  4.0   7.4  0.538  ...  10.8   
1   Adrian Branch  SF   23  LAL  32   0   6.8  1.5   3.0  0.500  ...   4.3   
2  Billy Thompson  SF   23  LAL  59   0  12.9  2.4   4.4  0.544  ...   5.6   
3     Byron Scott  SG   25  LAL  82  82  33.3  6.8  13.8  0.489  ...  17.0   
4    James Worthy  SF   25  LAL  82  82  34.4  7.9  14.7  0.539  ...  19.4   

   Year  Pts Won  Pts Max  Share                Team  Overall  Wins  Losses  \
0  1987      0.0    780.0    0.0  Los Angeles Lakers    65-17    65      17   
1  1987      0.0    780.0    0.0  Los Angeles Lakers    65-17    65      17   
2  1987      0.0    780.0    0.0  Los Angeles Lakers    65-17    65      17   
3  1987      0.0    780.0    0.0  Los Angeles Lakers    65-17    65      17   
4  1987      0.0    780.0    0.0  Los Angeles Lakers    65-17    65      17   

        W%  
0  0.79268  
1  0.79268  
2  0.79268  
3  0.79268  
4  0.79268  

[5 rows x 38 columns]

In [256]:
all_stats.columns

Index(['Player', 'Pos', 'Age', 'Tm', 'G', 'GS', 'MP', 'FG', 'FGA', 'FG%', '3P',
       '3PA', '3P%', '2P', '2PA', '2P%', 'eFG%', 'FT', 'FTA', 'FT%', 'ORB',
       'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS', 'Year',
       'Pts Won', 'Pts Max', 'Share', 'Team', 'Overall', 'Wins', 'Losses',
       'W%'],
      dtype='object')

In [258]:
#Delete overall columns since we have wins and losses
del all_stats['Overall']

In [259]:
all_stats.dtypes

Player      object
Pos         object
Age          int64
Tm          object
G            int64
GS           int64
MP         float64
FG         float64
FGA        float64
FG%        float64
3P         float64
3PA        float64
3P%        float64
2P         float64
2PA        float64
2P%        float64
eFG%       float64
FT         float64
FTA        float64
FT%        float64
ORB        float64
DRB        float64
TRB        float64
AST        float64
STL        float64
BLK        float64
TOV        float64
PF         float64
PTS        float64
Year         int64
Pts Won    float64
Pts Max    float64
Share      float64
Team        object
Wins         int64
Losses       int64
W%         float64
dtype: object

In [260]:
#Conver final data set to a dataframe
all_stats.to_csv('all_stats.csv')

Now we have a dataframe with all of the data we need to create a model. I will now create a notebook to do some data exploration before we build the model.